// twilio-run/dist/config/deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const deploy_1 = require("../commands/deploy");
const utils_1 = require("../commands/utils");
const utils_2 = require("../serverless-api/utils");
const global_1 = require("./global");
const utils_3 = require("./utils");
const mergeFlagsAndConfig_1 = require("./utils/mergeFlagsAndConfig");
function getConfigFromFlags(flags, externalCliOptions) {
    return __awaiter(this, void 0, void 0, function* () {
        let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
        flags.cwd = cwd;
        if (typeof flags.functionsEnv !== 'undefined') {
            utils_1.deprecateFunctionsEnv();
            if (!flags.environment) {
                flags.environment = flags.functionsEnv;
            }
            delete flags.functionsEnv;
        }
        const configFlags = global_1.readSpecializedConfig(cwd, flags.config, 'deployConfig', {
            projectId: flags.accountSid ||
                (externalCliOptions && externalCliOptions.accountSid) ||
                undefined,
            environmentSuffix: flags.environment,
        });
        flags = mergeFlagsAndConfig_1.mergeFlagsAndConfig(configFlags, flags, deploy_1.cliInfo);
        cwd = flags.cwd || cwd;
        if (flags.production) {
            flags.environment = '';
        }
        const { accountSid, authToken } = yield utils_3.getCredentialsFromFlags(flags, externalCliOptions);
        const { localEnv, envPath } = yield utils_3.readLocalEnvFile(flags);
        const serviceSid = flags.serviceSid ||
            (yield utils_2.getFunctionServiceSid(cwd, flags.config, 'deployConfig', flags.accountSid && flags.accountSid.startsWith('AC')
                ? flags.accountSid
                : externalCliOptions && externalCliOptions.accountSid));
        const pkgJson = yield utils_3.readPackageJsonContent(flags);
        const env = Object.assign({}, localEnv);
        for (let key of Object.keys(env)) {
            const val = env[key];
            if (typeof val === 'string' && val.length === 0) {
                delete env[key];
            }
        }
        delete env.ACCOUNT_SID;
        delete env.AUTH_TOKEN;
        let serviceName = yield utils_3.getServiceNameFromFlags(flags);
        if (!serviceName) {
            throw new Error('Please pass --service-name or add a "name" field to your package.json');
        }
        return {
            cwd,
            envPath,
            accountSid,
            authToken,
            env,
            serviceSid,
            pkgJson,
            overrideExistingService: flags.overrideExistingProject,
            force: flags.force,
            serviceName,
            functionsEnv: flags.environment,
            functionsFolderName: flags.functionsFolder,
            assetsFolderName: flags.assetsFolder,
            noAssets: !flags.assets,
            noFunctions: !flags.functions,
        };
    });
}
exports.getConfigFromFlags = getConfigFromFlags;
