// twilio-run/dist/config/list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const list_1 = require("../commands/list");
const utils_1 = require("../serverless-api/utils");
const global_1 = require("./global");
const utils_2 = require("./utils");
const mergeFlagsAndConfig_1 = require("./utils/mergeFlagsAndConfig");
function trim(str) {
    return str.trim();
}
function getConfigFromFlags(flags, externalCliOptions) {
    return __awaiter(this, void 0, void 0, function* () {
        let cwd = flags.cwd ? path_1.default.resolve(flags.cwd) : process.cwd();
        flags.cwd = cwd;
        const configFlags = global_1.readSpecializedConfig(cwd, flags.config, 'listConfig', {
            projectId: flags.accountSid ||
                (externalCliOptions && externalCliOptions.accountSid) ||
                undefined,
            environmentSuffix: flags.environment,
        });
        flags = mergeFlagsAndConfig_1.mergeFlagsAndConfig(configFlags, flags, list_1.cliInfo);
        cwd = flags.cwd || cwd;
        const { accountSid, authToken } = yield utils_2.getCredentialsFromFlags(flags, externalCliOptions);
        const serviceSid = flags.serviceSid ||
            (yield utils_1.getFunctionServiceSid(cwd, flags.config, 'listConfig'));
        let serviceName = yield utils_2.getServiceNameFromFlags(flags);
        const types = flags.types.split(',').map(trim);
        return {
            cwd,
            accountSid,
            authToken,
            serviceSid,
            serviceName,
            environment: flags.environment,
            properties: flags.properties
                ? flags.properties.split(',').map(x => x.trim())
                : undefined,
            extendedOutput: flags.extendedOutput,
            types,
        };
    });
}
exports.getConfigFromFlags = getConfigFromFlags;
