// twilio-run/dist/runtime/internal/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../../utils/logger");
const debug = logger_1.getDebugFunction('twilio-run:response');
class Response {
    constructor() {
        this.body = undefined;
        this.statusCode = 200;
        this.headers = {};
    }
    setStatusCode(statusCode) {
        debug('Setting status code to %d', statusCode);
        this.statusCode = statusCode;
    }
    setBody(body) {
        debug('Setting response body to %o', body);
        this.body = body;
    }
    setHeaders(headersObject) {
        debug('Setting headers to: %P', headersObject);
        if (typeof headersObject !== 'object') {
            return;
        }
        this.headers = headersObject;
    }
    appendHeader(key, value) {
        debug('Appending header for %s', key, value);
        this.headers = this.headers || {};
        this.headers[key] = value;
    }
    applyToExpressResponse(res) {
        debug('Setting values on response: %P', {
            statusCode: this.statusCode,
            headers: this.headers,
            body: this.body,
        });
        res.status(this.statusCode);
        res.set(this.headers);
        res.send(this.body);
    }
}
exports.Response = Response;
