// @twilio-labs/serverless-api/dist/api.js
"use strict";
/** @module @twilio-labs/serverless-api */
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const internals = __importStar(require("./api/index"));
const fs = __importStar(require("./utils/fs"));
const _utils = __importStar(require("./utils/index"));
exports.fsHelpers = fs;
exports.api = internals;
exports.utils = _utils;
var builds_1 = require("./api/builds");
exports.activateBuild = builds_1.activateBuild;
exports.triggerBuild = builds_1.triggerBuild;
exports.waitForSuccessfulBuild = builds_1.waitForSuccessfulBuild;
var dependencies_1 = require("./api/dependencies");
exports.getDependencies = dependencies_1.getDependencies;
var functions_1 = require("./api/functions");
exports.getOrCreateFunctionResources = functions_1.getOrCreateFunctionResources;
exports.uploadFunction = functions_1.uploadFunction;
var variables_1 = require("./api/variables");
exports.setEnvironmentVariables = variables_1.setEnvironmentVariables;
